pkg/server/server.go

```diff
@@ -310,20 +314,91 @@ func (s *Server) Run(ctx context.Context) error {
 		if !ok {
 			responsewriters.ErrorNegotiated(
 				apierrors.NewInternalError(fmt.Errorf("no RequestInfo found in the context")),
-				// TODO use the right Codecs
+				// TODO is this the right Codecs?
 				scheme.Codecs, schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}, res.ResponseWriter, req.Request,
 			)
 			return
 		}
-		if requestInfo.APIGroup == "" {
+
+		// If it's not the core group, pass through
+		if requestInfo.APIGroup != "" {
+			chain.ProcessFilter(req, res)
+			return
+		}
+
+		// Special handling of the core ("") group
+
+		if requestInfo.IsResourceRequest {
+			// This is a CRUD request for somethig like pods. Try to see if there is a CRD for the resource. If so, let the CRD
+			// server handle it.
 			crdName := requestInfo.Resource + ".core"
 			_, err := serverChain.CustomResourceDefinitions.Informers.Apiextensions().V1().CustomResourceDefinitions().Lister().GetWithContext(ctx, crdName)
 			if err == nil {
 				serverChain.CustomResourceDefinitions.GenericAPIServer.Handler.NonGoRestfulMux.ServeHTTP(res.ResponseWriter, req.Request)
 				return
 			}
+		} else if req.Request.URL.Path == "/api/v1" || req.Request.URL.Path == "/api/v1/" {
+			// This is a discovery request. We may need to combine discovery from the GenericControlPlane (which has built-in v1 types)
+			// and CRDs, if there are any v1 CRDs.
+
+			// If we are retrieving the GenericControlPlane's v1 APIResources, let that request through.
+			const passthroughHeader = "X-Kcp-Api-V1-Discovery-Passthrough"
+			if _, passthrough := req.Request.Header[passthroughHeader]; passthrough {
```

Comment:
who sets passthroughHeader ? could we share the constant with the code that sets this header?

Reply:
just 20 lines below I guess. Am still not sure why we get a recursion here (and have to protect against with this header).

Reply:
It was overflowing stack w/o this. Let me debug & figure out what's going on.

Reply:
am curious to see a backtrace of the recursion.

Reply:
The recursion happens because this entire function is a filter on the GoRestfulContainer, so this same filter function is invoked below when we call `serverChain.GenericControlPlane.GenericAPIServer.Handler.GoRestfulContainer.ServeHTTP`.

Reply:
And this required to be a filter function because the initial /api/v1 request is ultimately handled by:

https://github.com/kubernetes/kubernetes/blob/c16b2afc1d3c32462f068ea08cdc4791bd97b947/staging/src/k8s.io/apiserver/pkg/server/handler.go#L136